File: web/frontend/src/api/bands.ts
import type { ModelInfo } from './models'

/**
 * Confidence bands on the displayed SIF+ABTT similarity.
 *
 * THE BACKEND OWNS THESE NUMBERS (web/bands.py). They are not merely
 * presentational: the server uses `no_match` to decide whether a
 * reviewer-created directory has left `awaiting_match`, so a hardcoded copy
 * here would be a second source of truth able to drift from the one the API
 * answers with. `GET /api/models` therefore carries a `confidence_bands`
 * object on every entry and `bandsFrom` reads it.
 *
 * The literals below are the pre-flight fallback only, for the render that
 * happens before the first `/api/models` response lands. Keep them equal to
 * web/bands.py.
 */
export interface ConfidenceBands {
  /** Below this: overt "potentially no match", new-directory CTA is default. */
  no_match: number
  /** At or above this: calm "likely match - verify" framing. */
  verify: number
}

export const FALLBACK_BANDS: ConfidenceBands = { no_match: 0.5, verify: 0.7 }

/** The deployment's bands, or the fallback before the models call resolves. */
export function bandsFrom(models: ModelInfo[] | null | undefined): ConfidenceBands {
  return models?.[0]?.confidence_bands ?? FALLBACK_BANDS
}

export type ConfidenceLevel = 'no_match' | 'review' | 'likely'

/** Which band a score falls in. Boundaries are inclusive at the low end. */
export function bandOf(score: number, bands: ConfidenceBands): ConfidenceLevel {
  if (score < bands.no_match) return 'no_match'
  if (score < bands.verify) return 'review'
  return 'likely'
}

File: web/bands.py
"""Confidence bands on the displayed similarity.

One place, deliberately. Two features read the same numbers:

* the reviewer UI (issue #94) paints a candidate list red below ``NO_MATCH``,
  cautious between the bands, and calm above ``VERIFY``;
* the reviewer-directory loop (issue #95) calls a reviewer-created directory
  *matched* once some other query scores at least ``NO_MATCH`` against it.

The two must agree: a directory the backend calls matched has to be one the UI
paints as a match, or the badge and the card contradict each other on the same
screen.

WHO OWNS THE NUMBER. The backend, and the frontend reads it. The band is not
merely presentational here -- the server uses it to decide a persisted
directory's status -- so a hardcoded copy in TypeScript would be a second
source of truth that can silently drift from the one the API answers with.
``GET /api/models`` therefore carries a ``confidence_bands`` object on every
entry (alongside ``default_variant``, which is deployment-wide in the same
way), and the frontend's ``api/bands.ts`` reads it, falling back to these same
literals only before the first response lands.

CALIBRATION. Learned per-model tau under ``sif_abtt`` is 0.34-0.43 on the train
split, so 0.5 is deliberately conservative: a score below it is well under
every model's learned decision boundary, which is what makes an overt "no
match" treatment defensible.

These are similarity thresholds, not probabilities. Changing them changes what
reviewers are shown *and* which directories count as matched, so treat an edit
as a protocol change, not a tweak.
"""

from __future__ import annotations

#: Below this, the top candidate is flagged as "potentially no match" and the
#: new-directory CTA becomes the default action.
NO_MATCH_BAND: float = 0.5

#: At or above this, a candidate is framed as "likely match - verify".
VERIFY_BAND: float = 0.7

#: A reviewer-created directory leaves 'awaiting_match' once another query
#: scores at least this much against it. Same number as NO_MATCH_BAND by
#: definition: "matched" means "the UI would not paint this as a non-match".
REVIEWER_DIR_MATCH_BAND: float = NO_MATCH_BAND
